### ISSUE TITLE
Support tokenizing of the PHP 7.3 here/nowdoc syntax changes in older PHP versions

### ISSUE THREAD

**Issue by jrfnl** [squizlabs/PHP_CodeSniffer#2055]
PHP 7.3 will introduce flexible heredoc and nowdoc syntaxes. The PHPCS tokenizer currently does not handle these correctly.
This has been implemented already and should be testable via `nightly` or the PHP 7.3.0-alpha release which is available since this week.

> The closing marker for doc strings is no longer required to be followed by a semicolon or newline.
>  Additionally the closing marker may be indented, in which case the indentation will be stripped from all lines in the doc string.

Refs:
* https://github.com/php/php-src/blob/600cb4b46755a9170eaa32606528a6908bc98975/UPGRADING#L78
* https://wiki.php.net/rfc/flexible_heredoc_nowdoc_syntaxes

**Comment by gsherwood** [comment]
I ran PHPCS over some test cases from the RFC using PHP 7.3.0-dev and the tokenizing looks okay to me. Where are you seeing problems?

**Comment by jrfnl** [comment]
@gsherwood Sorry, I wasn't clear enough. It's not the tokenizing in PHP 7.3 which I'm worried about, it's the inconsistency in the tokenization when PHP 7.3 code is tokenized by PHPCS on any other PHP version.

This means that a sniff which - for instance - forbids the use of heredoc/nowdoc in function calls will break when running on PHP < 7.3 when confronted with PHP 7.3 heredoc/nowdoc syntax.

Example:
```php
<?php

$var = <<<EOD
	some text
	EOD . 'something else';

$var = my_function_call(<<<EOD
	some more text
	and more
	EOD,
	$param_too
);
```

would be tokenized like this on PHP < 7.3:
```
        Process token 0 on line 1 [col:1;len:5;lvl:0;]: T_OPEN_TAG => <?php\n
        Process token 1 on line 2 [col:1;len:0;lvl:0;]: T_WHITESPACE => \n
        Process token 2 on line 3 [col:1;len:4;lvl:0;]: T_VARIABLE => $var
        Process token 3 on line 3 [col:5;len:1;lvl:0;]: T_WHITESPACE =>
        Process token 4 on line 3 [col:6;len:1;lvl:0;]: T_EQUAL => =
        Process token 5 on line 3 [col:7;len:1;lvl:0;]: T_WHITESPACE =>
        Process token 6 on line 3 [col:8;len:6;lvl:0;]: T_STRING => <<<EOD\n
        Process token 7 on line 4 [col:1;len:10;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \tsome text\n
        Process token 8 on line 5 [col:1;len:24;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \tEOD . 'something else'
;\n
        Process token 9 on line 6 [col:1;len:0;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \n
        Process token 10 on line 7 [col:1;len:4;lvl:0;]: T_VARIABLE => $var
        Process token 11 on line 7 [col:5;len:26;lvl:0;]: T_ENCAPSED_AND_WHITESPACE =>  = my_function_call(<<<
EOD\n
        Process token 12 on line 8 [col:1;len:15;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \tsome more text\n
        Process token 13 on line 9 [col:1;len:9;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \tand more\n
        Process token 14 on line 10 [col:1;len:5;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \tEOD,\n
        Process token 15 on line 11 [col:1;len:1;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \t
        Process token 16 on line 11 [col:2;len:10;lvl:0;]: T_VARIABLE => $param_too
        Process token 17 on line 11 [col:12;len:0;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \n
        Process token 18 on line 12 [col:1;len:2;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => );\n
        Process token 19 on line 13 [col:1;len:0;lvl:0;]: T_ENCAPSED_AND_WHITESPACE => \n
```

**Comment by gsherwood** [comment]
Ah, I see what you mean.

This would require PHPCS to find and tokenize here/nowdocs itself in all PHP version < 7.3 when it currently relies on PHP to do this job. Not much fun 😢 

Presumably, it could still assume that if the end token is not indented and has a newline after it but was not tokenized as a here/nowdoc by PHP, it must have some other error and should be left as-is.

Maybe a little fun.

**Comment by jrfnl** [comment]
> Not much fun

I know, I'm sorry.

> Presumably, it could still assume that if the end token is not indented and has a newline after it but was not tokenized as a here/nowdoc by PHP, it must have some other error and should be left as-is.

That sounds about right.

I'm guessing that this will be quite some work anyway as most tokens after the PHP 7.3 heredoc/nowdoc token would be incorrectly tokenized on PHP 7.2 or lower... :unamused:

**Comment by gsherwood** [comment]
@jrfnl Any thoughts? If not, I'll close this off as I can't see how to achieve it.

**Comment by gsherwood** [comment]
@jrfnl I'm going to close this, but if you have any ideas or if it becomes a problem, we can revisit.

**Comment by jrfnl** [comment]
@gsherwood Well, it will be a problem for every single sniff which is run over a file containing the PHP 7.3 heredoc/nowdoc syntax when PHPCS is run on PHP < 7.3, but as concluded above, that looks to be an insurmountable one.

There is however a related issue which *may* be solvable.

In PHP 7.3, using the closing marker at the start of a line within the body of the heredoc/nowdoc is no longer allowed and will cause parse errors.
```php
<?php
echo <<<END
something
END{$var}
END;
```

The tokenizer in PHP itself gets confused over this code and will mark the first `END` as the `T_END_HEREDOC` and identify the second `END` as a `T_STRING`.

PHCPS on PHP 7.3 will tokenize the above example as:
```
0 :: L001 :: C1 :: T_OPEN_TAG :: (5) :: <?php

1 :: L002 :: C1 :: T_ECHO :: (4) :: echo
2 :: L002 :: C5 :: T_WHITESPACE :: (1) ::
3 :: L002 :: C6 :: T_START_HEREDOC :: (6) :: <<<END

4 :: L003 :: C1 :: T_HEREDOC :: (9) :: something

5 :: L004 :: C1 :: T_END_HEREDOC :: (3) :: END
6 :: L004 :: C4 :: T_OPEN_CURLY_BRACKET :: (1) :: {
7 :: L004 :: C5 :: T_VARIABLE :: (4) :: $var
8 :: L004 :: C9 :: T_CLOSE_CURLY_BRACKET :: (1) :: }
9 :: L004 :: C10 :: T_WHITESPACE :: (0) ::

10 :: L005 :: C1 :: T_STRING :: (3) :: END
11 :: L005 :: C4 :: T_SEMICOLON :: (1) :: ;
12 :: L005 :: C5 :: T_WHITESPACE :: (0) ::
```

I imagine it might just be possible to correct for this ? and to change this to the same tokenizer output as you would get when running PHPCS against the same code on PHP 7.2 or lower:
```
0 :: L001 :: C1 :: T_OPEN_TAG :: (5) :: <?php

1 :: L002 :: C1 :: T_ECHO :: (4) :: echo
2 :: L002 :: C5 :: T_WHITESPACE :: (1) ::
3 :: L002 :: C6 :: T_START_HEREDOC :: (6) :: <<<END

4 :: L003 :: C1 :: T_HEREDOC :: (9) :: something

5 :: L004 :: C1 :: T_HEREDOC :: (9) :: END{$var}

6 :: L005 :: C1 :: T_END_HEREDOC :: (3) :: END
7 :: L005 :: C4 :: T_SEMICOLON :: (1) :: ;
8 :: L005 :: C5 :: T_WHITESPACE :: (0) ::
```

**Comment by jrfnl** [comment]
> That's probably a bug worth reporting to PHP core regardless of what PHPCS does.

@gsherwood I don't think it's a bug, more a conscious BC-break as annotated in the RFC. As of PHP 7.3, the identifier of a heredoc/nowdoc is no longer allowed to be in the body of a heredoc/nowdoc at the start of a line. If it is, it will be a parse error, which is what is reflected in the tokenizer output.

https://wiki.php.net/rfc/flexible_heredoc_nowdoc_syntaxes#backward_incompatible_changes

**Comment by gsherwood** [comment]
> I don't think it's a bug, more a conscious BC-break as annotated in the RFC

Oh yeah sorry, the parse error was the bit I was missing there.
